File: packages/java/src/main/java/dev/kreuzberg/htmltomarkdown/ConversionResult.java
package dev.kreuzberg.htmltomarkdown;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.JsonNode;
import dev.kreuzberg.htmltomarkdown.metadata.HtmlMetadata;
import java.util.List;

/**
 * The primary result of {@code extract()}, containing all extracted content and structured data.
 *
 * @param content the converted Markdown string, or {@code null} in extraction-only mode
 * @param metadata extracted HTML metadata (title, links, images, etc.)
 * @param tables extracted tables with structured grid data
 * @param document structured document tree (raw JSON), or {@code null} if not requested
 * @param images extracted inline images (data URIs and SVGs)
 * @param warnings non-fatal processing warnings
 */
public record ConversionResult(
    @JsonProperty("content") String content,
    @JsonProperty("metadata") HtmlMetadata metadata,
    @JsonProperty("tables") List<ExtractTable> tables,
    @JsonProperty("document") JsonNode document,
    @JsonProperty("images") List<InlineImage> images,
    @JsonProperty("warnings") List<ProcessingWarning> warnings) { }

File: packages/csharp/HtmlToMarkdown/Metadata/ConversionResult.cs
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace HtmlToMarkdown.Metadata;

/// <summary>
/// The primary result of <c>Extract()</c>, containing all extracted content and structured data.
/// </summary>
public record ConversionResult
{
    /// <summary>
    /// The converted Markdown string, or <c>null</c> in extraction-only mode.
    /// </summary>
    [JsonPropertyName("content")]
    public string? Content { get; init; }

    /// <summary>
    /// Extracted HTML metadata (title, links, images, etc.).
    /// </summary>
    [JsonPropertyName("metadata")]
    public HtmlMetadata? Metadata { get; init; }

    /// <summary>
    /// Extracted tables with structured grid data.
    /// </summary>
    [JsonPropertyName("tables")]
    public List<ExtractTable> Tables { get; init; } = [];

    /// <summary>
    /// Structured document tree (raw JSON), or <c>null</c> if not requested.
    /// </summary>
    [JsonPropertyName("document")]
    public JsonElement? Document { get; init; }

    /// <summary>
    /// Extracted inline images (data URIs and SVGs).
    /// </summary>
    [JsonPropertyName("images")]
    public List<InlineImage> Images { get; init; } = [];

    /// <summary>
    /// Non-fatal processing warnings.
    /// </summary>
    [JsonPropertyName("warnings")]
    public List<ProcessingWarning> Warnings { get; init; } = [];
}
